### PR TITLE
feat: Zoe use watchPromise() to wait for contract finish

### LINKED ISSUE #8263: zoe will be confused by vat-vat-admin upgrade cancelling the 'done' promise
## What is the Problem Being Solved?

When a parent vat uses `E(vatAdminService).createVat()` to create a new vat, one thing it gets back is an `adminNode`: the "control facet" for the new vat. If you call `E(adminNode).done()`, the result Promise you get back will not generally fire until the new vat is upgraded.

However, vat-vat-admin could get upgraded, and all Promises decided by an upgraded vat are cancelled. This results in a distinctive rejection of the promise (the rejection value is an object, with `{ name: 'vatUpgraded' }` and some other properties), which can be recognized by the [`isUpgradeDisconnection`](https://github.com/Agoric/agoric-sdk/blob/2ffcef2e74f8837a348736edc1839b7c0f40a0c7/packages/internal/src/upgrade-api.js#L36) function from `@agoric/internal`.

While planning changes that would require `vat-vat-admin` to be upgraded, we realized that the currently-deployed Zoe will probably react to a vat-vat-admin upgrade as if the child vat had terminated, because the code that follows the `.done()` promise does not use `isUpgradeDisconnection`:

https://github.com/Agoric/agoric-sdk/blob/2ffcef2e74f8837a348736edc1839b7c0f40a0c7/packages/zoe/src/zoeService/startInstance.js#L283-L289

## Description of the Design

The zoe code needs to examine the rejection result using `isUpgradeDisconnection(reason)`. If `true`, then vat-vat-admin was upgraded, and zoe needs to fetch a new `done` promise (by calling `E(adminNode).done()`) and attach the same sort of callback. The original uses `E.when()` so I'm not quite sure of the sequence, but if it were using `.then`, the fixed version might look like:

```js
function reallyDone(adminNode) {
  return E(adminNode).done()
    .catch(err => isUpgradeDisconnection(err) ? reallyDone(adminNode) : throw(err));
 }

reallyDone(adminNode)
   .then(completion => instanceAdmin.exitAllSeats(completion),
         reason => instanceAdmin.failAllSeats(reason));
```

## Security Considerations

I know we use `E.when` as a guard against malicious promise-like objects with a fake `.then` method, and I wouldn't want to reduce our defensiveness by adding a `.then`, please adapt the code above to the local norms.

Contract vats which call `vatPowers.exitVatWithFailure(makeUpgradeDisconnection())` could confuse this code into an infinite loop, or maybe not, depending upon what `adminNode.done()` does on a terminated vat: I think the second call might trigger a different kind of error.

## Test Plan

We need Zoe unit tests that simulate the `done()` promise being abandoned, to ensure that Zoe will react properly. We also need Zoe upgrade tests to ensure that Zoe re-acquires the promise after an upgrade, or that its use of a DurablePromiseWatcher (see below) is doing the right thing.

## Upgrade Considerations

The deployed Zoe has the old code, and will react the bad way unless/until Zoe is upgraded. So we must defer any upgrades of vat-vat-admin until after a zoe upgrade that fixes this issue.

The zoe code that calls `E(adminNode).done()` lives in `startInstance`, which would not be called again during a Zoe upgrade. Unless we add something that re-runs `E(adminNode).done()` on all existing contract instances during a Zoe upgrade, I think a Zoe upgrade will basically forget the old handler, so subsequent contract vat terminations will just be ignored. That's probably bad for cleanup.

It may be a good idea to have Zoe use `globalThis.VatData.providePromiseWatcher()` to establish a durable watcher on the `done` Promise. That would remove the need to re-fetch a `done` promise when *zoe* is upgraded, however we would still need the `isUpgradedDisconnection()` check to guard against *vat-vat-admin* being upgraded.

cc @Chris-Hibbert to think about the Zoe changes
cc @mhofman to help me remember that we cannot upgrade vat-vat-admin until a Zoe fix+upgrade is deployed

### LINKED ISSUE #8265: robust handling of vat termination by Zoe across Zoe upgrades
See: #8263

## What is the Problem Being Solved?

Zoe queues up handling of vat termination in local promises which would be smashed by upgrade of the Zoe vat.

## Description of the Design

Handle (and test handling of) resetting those handlers when Zoe upgrades.

## Security Considerations

Any guarantees about final processing or notification should survive upgrade.

## Scaling Considerations

Not a huge deal. Has to be managed once per contract.

## Test Plan

use upgrade tests.

## Upgrade Considerations

Yep, that's what it's about.

### LINKED ISSUE #8387: zoe will stop tracking contract vat termination when vat-zoe is upgraded
As mentioned in https://github.com/Agoric/agoric-sdk/issues/8263#issuecomment-1696233044 , while investigating outstanding mainnet promises, we concluded that Zoe will fail to re-subscribe to contract-vat `done()` promises if/when vat-zoe is upgraded.

Zoe uses `vatAdminSvc` to create a new vat for each contract instance, which returns a per-vat `adminNode` that serves as a control facet. Zoe can use this to terminate the vat or trigger an upgrade. In addition, calling `E(adminNode).done()` will return a Promise that doesn't fire until the vat is terminated. The `done()` promise fulfills if the vat exits "happy" (if it self-terminates with `vatPowers.exitVat()`), and rejects if it exits "sad" (`vatPowers.exitVatWithFailure()`, `adminNode.terminateWithFailure()`, or the vat exceeds its meter or has some fatal error).

Zoe currently uses an ephemeral `E.when()` inside `startInstance` to attach a call to `exitAllSeats` to this promise:

https://github.com/Agoric/agoric-sdk/blob/2ffcef2e74f8837a348736edc1839b7c0f40a0c7/packages/zoe/src/zoeService/startInstance.js#L283-L289

which means that if/when vat-zoe is restarted or upgraded, the callback will be lost. The zoe vat will still be subscribed to the promise (the kernel will deliver a `dispatch.notify`), but the new zoe incarnation will not have any callbacks registered to fire, so it will ignore the resolution.

As a result, when vat-zoe is upgraded, it will stop tracking that `done()` promise. So if any contract vats terminate, zoe will not clean up after them. Depending upon the exit conditions on each offer, that might make it difficult to close out a seat.

One possible fix is for Zoe to use globalThis.VatData.providePromiseWatcher() to establish a durable watcher on the done Promise. That would remove the need to re-fetch a done promise when zoe is upgraded (however we would still need the isUpgradedDisconnection() check to guard against vat-vat-admin being upgraded). However, that won't help with the existing `done()` promises currently tracked by the deployed zoe.

Another fix, which *would* help with the current deployment, would be for Zoe's `buildRootObject` to scan its durable storage for the `adminNode`s of all active contract vats, and call `done()` on each of them, to re-establish the callbacks.

If we wanted to combine the two techniques, zoe's per-contract-vat table should have a flag to track whether we have a durable promise watcher registered or not. At startup (`buildRootObject`), walk the table, and for every vat that lacks the flag, use `providePromiseWatcher()`, and set the flag. In `startInstance`, when starting a new contract vat, use `providePromiseWatcher()` and also set the flag.

### LINKED ISSUE #8915: Write a Zoe coreEval proposal for upgrade-16
## What is the Problem Being Solved?

**Upgrade-15** should upgrade Zoe to the code currently on master. #8868 details some tests of Zoe.

## Description of the Design

Pretty much a copy of replace-zoe.js. It should upgrade Zoe and update the bundleCap for ZCF.

## Security Considerations

N/A

## Scaling Considerations

N/A

## Test Plan

See #8868.

## Upgrade Considerations

yes.

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
fixes: #8387
fixes: #8265
fixes: #8263
fixes: #8915 

## Description

Zoe wants to exit all seats when the contract completes. It currently attempts to do that with an ephemeral promise, which will break when Zoe itself is upgraded. This uses durable watched promises to surmount that problem.

### Security Considerations

Zoe's durability guarantees, and ability to maintain payout liveness when a contract crashes.

### Scaling Considerations

Not a major issue.

### Documentation Considerations

N/A

### Testing Considerations

There are two tests of upgrade in zoe (test-zoe-upgrade and upgrade-covered-call). These tests invoke the new `onFulfilled` handler. 

This PR updates `test-zoe-upgrade` to include a restart (null upgrade) of the vat-admin vat.  With that and a fix ([581e9ad8](https://github.com/Agoric/agoric-sdk/pull/8453/commits/581e9ad87a586feb905312fa38e6d2f95067a887)) to the test, the old code fails the upgrade, and the new code passes.

### Upgrade Considerations

It's all about upgrade.

### CODE REVIEW

**Comment by turadg** [comment]
https://github.com/Agoric/agoric-3-proposals/pull/18 tests upgrading agoric-3 with this version of Zoe.

The upgrade is failing due to: "durable Kind stateShape mismatch (body)"

```
#102 37.46 2023-11-08T23:27:35.351Z SwingSet: vat: v1: evaluateBundleCap { manifestBundleRef: { bundleID: 'b1-a3dc21c56535b4bebdf8c8413650eec2b37a6d9f5669b27f8c571601a6f641ffae2b925dd7e3d67c6953a1f8122c8ba814305e034c9087056816a6f05dd41069' }, exportedGetManifest: 'getManifestForUpgradingZoe', vatAdminSvc: Promise [Promise] {} }
#102 37.53 2023-11-08T23:27:35.421Z SwingSet: vat: v1: execute { exportedGetManifest: 'getManifestForUpgradingZoe', behaviors: [ 'getManifestForUpgradingZoe', 'upgradeZoe' ] }
#102 37.53 2023-11-08T23:27:35.422Z SwingSet: vat: v1: coreProposal: upgradeZoe
#102 37.54 2023-11-08T23:27:35.432Z SwingSet: vat: v1: ZOE BUNDLE ID:  b1-31f084378c798adc3018ac20670b43774adeb2c17e67c5676909a7ae7a806cb3c23cb671252d829a65f12ca052be8ef796dd9815d651b9e0372cfb8699e6e939
#102 38.20 block produced
#102 38.24 Waiting for proposal 7 to pass (status=PROPOSAL_STATUS_VOTING_PERIOD)
#102 38.24 1
#102 38.50 2023-11-08T23:27:36.391Z SwingSet: xsnap: v9: error during vat dispatch() of startVat,[object Object] (Error#1)
#102 38.50 2023-11-08T23:27:36.391Z SwingSet: xsnap: v9: Error#1: durable Kind stateShape mismatch (body)
#102 38.50 2023-11-08T23:27:36.391Z SwingSet: xsnap: v9: Error: durable Kind stateShape mismatch (body)
#102 38.50  at construct ()
#102 38.50  at Error (/bundled-source/.../node_modules/ses/src/error/tame-error-constructor.js:56)
#102 38.50  at makeError (/bundled-source/.../node_modules/ses/src/error/assert.js:243)
#102 38.50  at fail (/bundled-source/.../node_modules/ses/src/error/assert.js:357)
#102 38.50  at insistSameCapData (/bundled-source/.../packages/swingset-liveslots/src/virtualObjectManager.js:279)
#102 38.50  at defineKindInternal (/bundled-source/.../packages/swingset-liveslots/src/virtualObjectManager.js:799)
#102 38.50  at defineDurableKind (/bundled-source/.../packages/swingset-liveslots/src/virtualObjectManager.js:1155)
#102 38.50  at makeStartInstance (.../zoe/src/zoeService/startInstance.js:213)
#102 38.50  at makeDurableZoeKit (.../zoe/src/zoeService/zoe.js:154)
#102 38.50  at buildRootObject (.../vats/src/vat-zoe.js:21)
#102 38.50  at startVat (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1476)
#102 38.50  at ()
#102 38.50  at ()
#102 38.50
#102 38.50 2023-11-08T23:27:36.391Z SwingSet: xsnap: v9: UnhandledPromiseRejectionWarning: (Error#1)
#102 38.51 2023-11-08T23:27:36.395Z SwingSet: kernel: WARNING: vat v9 failed to upgrade from incarnation 0 (startVat)
```

**Comment by turadg** [comment]
> The upgrade is failing due to: "durable Kind stateShape mismatch (body)"

Fixed by reverting the string in the `M.remotable` within `InstanceAdminShape`. I amended and force-pushed. Now it just has a comment.

**Comment by Chris-Hibbert** [comment]
PTAL, @erights @warner @turadg 
This is the primary PR on the path to upgrading Zoe. Once it's approved for the branch master, I'll make a rebase to the dev-branch for staging for release to the chain.

File: packages/zoe/src/zoeService/startInstance.js [hunk 1]
```diff
@@ -53,10 +60,44 @@ export const makeStartInstance = (
...
   });
 
+  const watcher = prepareExo(
+    zoeBaggage,
+    'InstanceCompletionWatcher',
+    M.interface('InstanceCompletionWatcher', {
+      onFulfilled: M.call(
+        M.any(),
+        InstanceAdminShape,
+        M.remotable('adminNode'),
```
Comment (turadg) [inline packages/zoe/src/zoeService/startInstance.js:77]: This param isn't used. I thought we could remove it, but then remembered that in the latest Endo, Exo methods fail if they are passed extra arguments. Filing for future work: https://github.com/Agoric/agoric-sdk/issues/8709

**Review by turadg — APPROVED** [review]

**Comment by Chris-Hibbert** [comment]
I'm getting conflicting opinions from tslint. If I have the following `ts-expect-error` in `startInstance.js`
```
          // @ts-expect-error XXX reason isn't necessarily an Error
          instanceAdmin.failAllSeats(reason);
```

then `yarn lint` from packages/zoe is happy, but when run from packages/vats, it complains **../zoe/src/zoeService/startInstance.js:92:11 - error TS2578: Unused '@ts-expect-error' directive.**

If I leave out the `ts-expect-error`, then lint from .../vats is happy, but from .../zoe is complains **src/zoeService/startInstance.js:92:38 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Error'.**

Somehow CI runs both checks. Is there a way to make them both happy?

**Pull request merged**